### ISSUE TITLE
URL vs String

### ISSUE THREAD

**Issue by rob-nash** [JohnSundell/Files#59]
Discussion

Would it be better to return instances of [URL](https://developer.apple.com/documentation/foundation/nsurl) wherever possible, instead of stringPath ?

> URL objects are the preferred way to refer to local files. Most objects that read data from or write data to a file have methods that accept an NSURL object instead of a pathname as the file reference. For example, you can get the contents of a local file URL as an NSString object using the init(contentsOf:encoding:) initializer, or as an NSData object using the init(contentsOf:options:) initializer.

> You can also use URLs for interapplication communication. In macOS, the NSWorkspace class provides the open(_:) method to open a location specified by a URL. Similarly, in iOS, the UIApplication class provides the openURL(_:) method.

> Additionally, you can use URLs when working with pasteboards, as described in NSURL Additions Reference (part of the AppKit framework).

> The NSURL class is “toll-free bridged” with its Core Foundation counterpart, CFURL. See Toll-Free Bridging for more information on toll-free bridging.

**Comment by clayellis** [comment]
After a deeper dive, looks like `FileManager` itself relies heavily on `String` paths which may be enough precedent to stick with `String`. I'd still be interested in hearing your thoughts, @JohnSundell.

**Comment by rob-nash** [comment]
Got this from a WWDC vid 2015 showing the power of URL's.

<img width="1735" alt="screen shot 2018-05-18 at 11 06 13" src="https://user-images.githubusercontent.com/14126999/40532098-46e0735c-5ff6-11e8-947a-79de8e593427.png">

<img width="1169" alt="screen shot 2018-05-18 at 11 06 48" src="https://user-images.githubusercontent.com/14126999/40532101-4987eeb4-5ff6-11e8-9424-3298abb4004d.png">

**Comment by clayellis** [comment]
Right, `URL` has a very robust API. I love using them. I wonder why [FileManager](https://developer.apple.com/documentation/foundation/filemanager) uses `String` everywhere.

**Comment by clayellis** [comment]
I've started a [WIP branch](https://github.com/clayellis/Files/tree/url) for the conversion. 

This is extremely preliminary branch with all tests passing. A few of the public APIs were broken (with plans to undo those breaking changes) but were fixed with a couple internal extensions (that we might want to consider making public).

**Comment by clayellis** [comment]
The [URL branch](https://github.com/clayellis/Files/tree/url) is starting to look better. Let's have a conversation about the tradeoffs before moving forward with a PR.

**Comment by rob-nash** [comment]
That makes sense @JohnSundell. We could provide public API that offers both? So you can pass in a URL or a string. Apple seems to acknowledge the usefulness of both approaches.

```
Bundle.main.url(forResource: "", withExtension: "") 
Bundle.main.path(forResource: "", ofType: "")
```

```
FileManager.default.moveItem(atPath: "", toPath: "")
FileManager.default.moveItem(at: URL(), to: URL()
```

**Comment by JohnSundell** [comment]
Yeah 👍

**Comment by rob-nash** [comment]
Hi @clayellis 

The url branch isn't stable at this very moment in time. The target is set to iOS 8 but the code is touching iOS 9 API.

I guess bumping the deployment target is the first noticeable trade-off.

I'll bump the deployment target locally, so that it compiles on my machine but I won't be making any contributions to your branch right now.

At first glance, it's looking good.

**Comment by clayellis** [comment]
@rob-nash I haven't tested targeting iOS, I've just been targeting macOS while playing around.

@JohnSundell yeah, that's what I've felt, too. Files is primarily meant for scripting and having to deal in URLs is a pain whereas passing in Strings is super convenient. I'm just wondering which route we want to take:
1. Expose [this URL extension](https://github.com/clayellis/Files/blob/0d0c43e5fe4cc9144bc3c262c8a352fc304cd558/Sources/Files.swift#L1154) publicly so that Strings can be passed wherever URLs are expected.
or
2. Hide that extension, but create counterpart methods, one for URL, one for String (i.e.: `File(url: URL)` & `File(path: String)`)

Which direction do you like more?

**Comment by clayellis** [comment]
@rob-nash I've made the branch backwards compatible with iOS 8 and tvOS 9 👍

**Comment by clayellis** [comment]
@JohnSundell is there a reason why the home directory is accessed [through environment variables](https://github.com/JohnSundell/Files/blob/06f95bd1bf4f8d5f50bc6bb30b808c253acd4c88/Sources/Files.swift#L1111) instead of `NSHomeDirectory()`? I feel like that may have been on purpose, but just want to make sure?

All of the tests pass when using `NSHomeDirectory()`.

**Comment by clayellis** [comment]
@rob-nash I've made the branch backwards compatible with iOS 8 and tvOS 9 👍

**Comment by rob-nash** [comment]
@clayellis that would explain it! I'm always wearing my iOS hat 🤠😄

**Comment by rob-nash** [comment]
Is it worth removing some of the boiler plate equatable implementations, given that [Swift now handles that](https://twitter.com/johnsundell/status/999742519205392389)?

https://github.com/JohnSundell/Files/blob/06f95bd1bf4f8d5f50bc6bb30b808c253acd4c88/Sources/Files.swift#L56
https://github.com/JohnSundell/Files/blob/06f95bd1bf4f8d5f50bc6bb30b808c253acd4c88/Sources/Files.swift#L98
etc

**Comment by clayellis** [comment]
@rob-nash Yeah, probably 👍
